### ISSUE TITLE
"ValueError: Cannot set phased before genotype is set"

### ISSUE THREAD

**Issue by iek** [whatshap/whatshap#356]
Hello, when I attempt to run whatshap, I get the following error message:

![Screen Shot 2022-02-22 at 8 08 04 PM](https://user-images.githubusercontent.com/12649783/155246148-af44a31d-0ba3-473b-b51d-60e2901fa789.png)

I am not sure how to proceed. Do you have suggestions? Thank you so much.

**Comment by Sj1925** [comment]
I’m having the same issue as well. Is there something wrong with the sorted bam file maybe?

**Comment by marcelm** [comment]
How did you install WhatsHap? pip or Conda? Which pysam versions are you using?

**Comment by iek** [comment]
Thank you for your response. I installed WhatsHap on Conda. I am using Pysam-0-.18.0 . I am now getting a new KeyError, however. 

![Screen Shot 2022-02-23 at 3 51 19 PM](https://user-images.githubusercontent.com/12649783/155406610-dc98fe7d-efcf-4356-bc3b-242bd3dd0859.png)

**Comment by marcelm** [comment]
What did you do differently compared to before? What is the command line that you are using? Can you perhaps share the data?

@Sj1925 It’s more likely an issue with the VCF/BCF.

**Comment by marcelm** [comment]
And if possible, please avoid pasting screenshots and just copy and paste the text into the issue. That makes it easier for me to work with. Use the little <> button above the text box to format it as code.

**Comment by iek** [comment]
For the first error, I used Minimap to map the Illumina reads. I then changed to BWA Mem and got the second "KeyError: invalid format". Do you have an email I could send our data to? And yes, absolutely sorry about that.

**Comment by marcelm** [comment]
Just click on my user name to see my profile; it shows my e-mail address.

**Comment by iek** [comment]
Sounds good. Thank you. I also mapped my long reads with Minimap.

**Comment by marcelm** [comment]
I was able to fix both the `KeyError: 'invalid FORMAT: GT'` and the `ValueError: Cannot set phased before genotype is set` problems, but I am getting a different error on @iek’s data, so I’m re-opening this for the moment.

**Comment by marcelm** [comment]
The new problem is that the `GQ` field in the VCF has type "Float" while it should be "Integer" according to the VCF standard.

I have now changed WhatsHap so that it is less strict and accepts GQ fields of type Float as well.

And finally, I was *still* not able to phase the data due to this problem:
```
VCF not ordered: Pf3D7_05_v3:1200269 appears before Pf3D7_05_v3:91
```
And this is indeed the case in the VCF file:
```
Pf3D7_04_v3     1200267 .       G       <*>     0       .       DP=0;END=1200267;MIN_DP=0       GQ:DP:MIN_DP:QR:RO:QA:AO        -0:0:0:0:0:0:0
Pf3D7_04_v3     1200268 .       G       <*>     0       .       DP=0;END=1200268;MIN_DP=0       GQ:DP:MIN_DP:QR:RO:QA:AO        -0:0:0:0:0:0:0
Pf3D7_05_v3     1200269 .       G       <*>     0       .       DP=0;END=1200490;MIN_DP=0       GQ:DP:MIN_DP:QR:RO:QA:AO        -0:0:0:0:0:0:0
Pf3D7_05_v3     91      .       A       <*>     0       .       DP=0;END=91;MIN_DP=0    GQ:DP:MIN_DP:QR:RO:QA:AO        -0:0:0:0:0:0:0
Pf3D7_05_v3     92      .       C       <*>     0       .       DP=0;END=92;MIN_DP=0    GQ:DP:MIN_DP:QR:RO:QA:AO        -0:0:0:0:0:0:0
```
Note that the first two rows are on chromosome `_04_v3` and the four following ones are on `_05_v3`. This looks very much like a bug in FreeBayes because it seems as if the third row should actually be on `_04_v3` as well.

It seems this issue is known, but was never fixed, see https://github.com/freebayes/freebayes/issues/561. I suggest you comment over there and try to get the developers to re-open the issue. Link to this issue here if you want. As a workaround, you could try to use `bcftools sort`. The record with the incorrect contig will still be there, but at least it will be sorted "correctly" so that WhatsHap doesn’t complain.

If you want to try the WhatsHap version with the bug fixes I made today, please read https://whatshap.readthedocs.io/en/latest/installation.html#installing-an-unreleased-development-version, which I have just updated.

Please let me know how it goes. If you have some use of the changes I made, I will make a new WhatsHap release soon.

**Comment by marcelm** [comment]
Another related FreeBayes issue: https://github.com/freebayes/freebayes/issues/639

**Comment by iek** [comment]
Hi Marcel, thank you very much. This fixed the issue for "whatshap phased". I just manually went through the VCF file to fix the errors that Freebayes introduced. I would love to see this implemented in a released version.

Unfortunately, I am trying to visualize the read blocks now with "whatshap haplotag" and am running into the same issue with KeyError: 'invalid FORMAT: GT' . Would it be possible to fix the same issue in "whatshap haplotag" as well?

`Found 1 sample(s) in input VCF
Found 0 sample(s) in BAM file
Traceback (most recent call last):
  File "/users/iek/anaconda/whatshap-tmp/bin/whatshap", line 8, in <module>
    sys.exit(main())
  File "/users/iek/anaconda/whatshap-tmp/lib/python3.10/site-packages/whatshap/__main__.py", line 64, in main
    module.main(args)
  File "/users/iek/anaconda/whatshap-tmp/lib/python3.10/site-packages/whatshap/cli/haplotag.py", line 622, in main
    run_haplotag(**vars(args))
  File "/users/iek/anaconda/whatshap-tmp/lib/python3.10/site-packages/whatshap/cli/haplotag.py", line 540, in run_haplotag
    variant_table = load_chromosome_variants(vcf_reader, chrom, regions)
  File "/users/iek/anaconda/whatshap-tmp/lib/python3.10/site-packages/whatshap/cli/haplotag.py", line 146, in load_chromosome_variants
    variant_table = vcf_reader.fetch_regions(chromosome, regions)
  File "/users/iek/anaconda/whatshap-tmp/lib/python3.10/site-packages/whatshap/vcf.py", line 424, in fetch_regions
    return self._process_single_chromosome(chromosome, records)
  File "/users/iek/anaconda/whatshap-tmp/lib/python3.10/site-packages/whatshap/vcf.py", line 505, in _process_single_chromosome
    p = extract_phase(call)
  File "/users/iek/anaconda/whatshap-tmp/lib/python3.10/site-packages/whatshap/vcf.py", line 449, in _extract_GT_PS_phase
    is_het = not all(x == call["GT"][0] for x in call["GT"])
  File "pysam/libcbcf.pyx", line 3478, in pysam.libcbcf.VariantRecordSample.__getitem__
  File "pysam/libcbcf.pyx", line 816, in pysam.libcbcf.bcf_format_get_value
KeyError: 'invalid FORMAT: GT'`

**Comment by marcelm** [comment]
This should also be fixed now (commit fc4c45e). Please run `conda activate whatshap-tmp` follewd by `pip install git+https://github.com/whatshap/whatshap` to update.

**Comment by iek** [comment]
Thank you so much for your time and help. The program ran successfully.
